Containerfile

```diff
@@ -1,47 +1,55 @@
 # hadolint global ignore=DL3007
-FROM registry.access.redhat.com/ubi9/ubi-minimal:latest
-LABEL maintainer="Red Hat"
 
+########################
+# PREPARE OUR BASE IMAGE
+########################
+FROM registry.access.redhat.com/ubi9/ubi-minimal:latest as base
+RUN microdnf -y install \
+    --setopt install_weak_deps=0 \
+    --nodocs \
+    git-core \
+    nodejs \
+    python3 \
+    && microdnf clean all
+
+######################
+# BUILD/INSTALL CACHI2
+######################
+FROM base as builder
 WORKDIR /src
+COPY . .
 RUN microdnf -y install \
     --setopt install_weak_deps=0 \
     --nodocs \
     gcc \
-    git-core \
     golang-bin \
     nodejs \
     npm \
-    python3 \
     python3-devel \
     python3-pip \
     python3-setuptools \
     && microdnf clean all
 
-COPY . .
+RUN python3 -m venv /venv && \
+    /venv/bin/pip install -r requirements.txt --no-deps --no-cache-dir --require-hashes && \
+    /venv/bin/pip install --no-cache-dir .
 
-RUN pip3 install -r requirements.txt --no-deps --no-cache-dir --require-hashes && \
-    pip3 install --no-cache-dir . && \
-    # the git folder is only needed to determine the package version
-    rm -rf .git
+##########################
+# ASSEMBLE THE FINAL IMAGE
+##########################
+FROM base
+LABEL maintainer="Red Hat"
 
-WORKDIR /src/js-deps
-RUN npm install && \
-    ln -s "${PWD}/node_modules/.bin/corepack" /usr/local/bin/corepack && \
-    corepack enable yarn && \
-    microdnf -y remove npm
+# copy Go SDKs from official Debian images, corepack from official Node.js Alpine
+COPY --from=docker.io/library/golang:1.20.0-bullseye /usr/local/go /usr/local/go/go1.20
```

Comment:
Do we build this often enough to need to worry about DockerHub rate limits?

Reply:
That's a very good question but because I don't have admin rights on the org and not seeing all the settings I wonder if caching could be used, i.e. cache the intermediary images indefinitely and just copy from them which would actually be awesome because it'd speed up even GH builds by a huge margin.

Reply:
I'd not be concerned about dockerlimits, we can provide docker credentials if needed

Reply:
Relying on docker credentials only exhibits the strong ties to a particular organization behind the project which IMO doesn't foster the OSS community principles, IIUC docker should have a way to cache the intermediary layers during a multi-stage build, although admittedly, everything I could find looked really hackish. Agreed with @MartinBasti that we should take care of the problem when we face it.

Reply:
you don't want to mess with layer caching, mainly when you are just installing RPMs and the particular line doesn't change, docker can reuse layer with old RPMs. It's safer to build image without relaying on layers.

And layers won't address the initial problem that you have to pull parent image first and also you want "latest" so you cannot just cache it, as docker would reuse that local tag forever.

We should be fine building this in GH, limits are quite generous.

Reply:
> you don't want to mess with layer caching, mainly when you are just installing RPMs and the particular line doesn't change, docker can reuse layer with old RPMs. It's safer to build image without relaying on layers.

FYI this only works for regular builds, with multi-stage there's more going on and the RPM layers tend to be rebuilt every time (not sure what the reason is I haven't investigated, maybe just ordering).